plugin/evm/acp176/acp176.go

```diff
@@ -0,0 +1,197 @@
+	return gas.Gas(gas.CalculatePrice(
+		MinTargetPerSecond,
+		s.TargetExcess,
+		TargetConversion,
+	))
+}
+
+// MaxCapacity returns the maximum possible accrued gas capacity.
+func (s *State) MaxCapacity() gas.Gas {
+	targetPerSecond := s.Target()
+	maxCapacity, err := safemath.Mul(targetToMaxCapacity, targetPerSecond)
+	if err != nil {
+		maxCapacity = math.MaxUint64
+	}
+	return maxCapacity
+}
+
+// BaseFee returns the current required fee per gas.
+//
+// BaseFee = MinBaseFee * e^(Excess / (Target() * TargetToPriceUpdateConversion))
+func (s *State) BaseFee() *big.Int {
+	target := s.Target()
+	priceUpdateConversion, err := safemath.Mul(TargetToPriceUpdateConversion, target)
+	if err != nil {
+		priceUpdateConversion = math.MaxUint64
+	}
+
+	bigExcess := new(big.Int).SetUint64(uint64(s.Gas.Excess))
+	bigPriceUpdateConversion := new(big.Int).SetUint64(uint64(priceUpdateConversion))
+	return fakeExponential(minBaseFee, bigExcess, bigPriceUpdateConversion)
+}
+
+// AdvanceTime increases the gas capacity and decreases the gas excess based on
+// the elapsed seconds.
+func (s *State) AdvanceTime(seconds uint64) {
+	targetPerSecond := s.Target()
+	maxPerSecond, err := safemath.Mul(TargetToMax, targetPerSecond)
+	if err != nil {
+		maxPerSecond = math.MaxUint64
+	}
+	maxCapacity, err := safemath.Mul(TimeToFillCapacity, maxPerSecond)
+	if err != nil {
+		maxCapacity = math.MaxUint64
+	}
+	s.Gas = s.Gas.AdvanceTime(
+		maxCapacity,
+		maxPerSecond,
+		targetPerSecond,
+		seconds,
+	)
+}
+
+// ConsumeGas decreases the gas capacity and increases the gas excess by
+// gasUsed + extraGasUsed. If the gas capacity is insufficient, an error is
+// returned.
+func (s *State) ConsumeGas(
+	gasUsed uint64,
+	extraGasUsed *big.Int,
+) error {
+	var err error
+	s.Gas, err = s.Gas.ConsumeGas(gas.Gas(gasUsed))
+	if err != nil {
+		return err
+	}
+	if extraGasUsed != nil {
+		if !extraGasUsed.IsUint64() {
+			return fmt.Errorf("%w: extraGasUsed (%d) exceeds MaxUint64",
+				gas.ErrInsufficientCapacity,
+				extraGasUsed,
+			)
+		}
+		s.Gas, err = s.Gas.ConsumeGas(gas.Gas(extraGasUsed.Uint64()))
+	}
+	return err
+}
+
+// UpdateTargetExcess updates the targetExcess to be as close as possible to the
+// desiredTargetExcess without exceeding the maximum targetExcess change.
+func (s *State) UpdateTargetExcess(desiredTargetExcess gas.Gas) {
+	previousTargetPerSecond := s.Target()
+	s.TargetExcess = targetExcess(s.TargetExcess, desiredTargetExcess)
+	newTargetPerSecond := s.Target()
+	s.Gas.Excess = scaleExcess(
+		s.Gas.Excess,
+		newTargetPerSecond,
+		previousTargetPerSecond,
+	)
+}
+
+// DesiredTargetExcess calculates the optimal desiredTargetExcess given the
+// desired target.
+//
+// This could be solved directly by calculating D * ln(desiredTarget / P) using
+// floating point math. However, it introduces inaccuracies. So, we use a binary
+// search to find the closest integer solution.
+func DesiredTargetExcess(desiredTarget gas.Gas) gas.Gas {
```

Comment:
IIUC, this calculates $q_{desired}$ as the first value that puts the gas target above the nodes set desired gas target.

It likely doesn't matter either way, but should we err on the side of going with the last value that puts the gas target below their desired level to ensure it never recommends exceeding it? Or potentially going with the value of $q$ that puts the gas target closest to the configured desired level rather than always rounding up.

Reply:
I felt like people would probably select round numbers. For example: `1.5M gps` or `2M gps`. If we go on the low side, then the actual targets could look like `1,499,998 gps` vs `1,500,001 gps` (or something). I felt like just visually, going slightly over looks nicer...

I don't think this actually matters as far as the protocol is concerned... Also we are able to perfectly calculate:
- 1,000,000
- 1,500,000
- 3,000,000
- 6,000,000
- 10,000,000
- 100,000,000

So for reasonable numbers the difference is nonexistent.